File: collector/pkg/component/consumer/exporter/cameraexporter/filewriter.go
package cameraexporter

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"os"
	"path"
	"path/filepath"
	"sort"
	"strconv"

	"github.com/Kindling-project/kindling/collector/pkg/component"
	"github.com/Kindling-project/kindling/collector/pkg/filepathhelper"
	"github.com/Kindling-project/kindling/collector/pkg/model"
	"github.com/Kindling-project/kindling/collector/pkg/model/constlabels"
	"github.com/Kindling-project/kindling/collector/pkg/model/constnames"
)

type fileWriter struct {
	config *fileConfig
	logger *component.TelemetryLogger
}

func newFileWriter(cfg *fileConfig, logger *component.TelemetryLogger) (*fileWriter, error) {
	// StoragePath must be an absolute path
	if !path.IsAbs(cfg.StoragePath) {
		return nil, fmt.Errorf("storage_path must be an absolute path")
	}
	// Check if there is the directory.
	// If the directory is not found, we create one
	if err := os.MkdirAll(cfg.StoragePath, os.ModePerm); err != nil {
		return nil, fmt.Errorf("failed to create fileWriter: %w", err)
	}
	return &fileWriter{
		config: cfg,
		logger: logger,
	}, nil
}

func (fw *fileWriter) write(group *model.DataGroup) {
	groupName := group.Name
	switch groupName {
	case constnames.SingleNetRequestMetricGroup:
		fw.writeTrace(group)
	case constnames.SpanEvent:
		fw.writeTrace(group)
	case constnames.CameraEventGroupName:
		fw.writeCpuEvents(group)
	}
}

func (fw *fileWriter) pidFilePath(workloadName string, podName string, containerName string, pid int64) string {
	dirName := workloadName + "_" + podName + "_" + containerName + "_" + strconv.FormatInt(pid, 10)
	return path.Join(fw.config.StoragePath, dirName)
}

func getFileName(protocol string, contentKey string, timestamp uint64, isServer bool) string {
	var isServerString string
	if isServer {
		isServerString = "true"
	} else {
		isServerString = "false"
	}
	encodedContent := base64.URLEncoding.EncodeToString([]byte(contentKey))
	return getDateString(int64(timestamp)) + "_" + protocol + "_" + encodedContent + "_" + isServerString
}

func (fw *fileWriter) writeTrace(group *model.DataGroup) {
	pathElements := filepathhelper.GetFilePathElements(group, group.Timestamp)
	// Create the directory first for saving its profile files.
	// If there has been such a directory, it will do nothing and return nil.
	baseDir := fw.pidFilePath(pathElements.WorkloadName, pathElements.PodName, pathElements.ContainerName, pathElements.Pid)
	if err := os.MkdirAll(baseDir, os.ModePerm); err != nil {
		fw.logger.Errorf("Failed to create pid directory: %v", err)
		return
	}
	// /$path/podName_containerName_pid/protocol_contentKey_timestamp_isServer
	fileName := getFileName(pathElements.Protocol, pathElements.ContentKey, pathElements.Timestamp, pathElements.IsServer) // Check whether we need to roll over the files
	err := fw.writeFile(baseDir, fileName, group)
	if err != nil {
		fw.logger.Errorf("Failed to write trace to file: %v", err)
		return
	}
}

func (fw *fileWriter) writeFile(baseDir string, fileName string, group *model.DataGroup) error {
	// Check whether the count of files is greater than MaxCount
	err := fw.rotateFiles(baseDir)
	if err != nil {
		fw.logger.Warnf("can't rotate files in %s: %v", baseDir, err)
	}
	filePath := filepath.Join(baseDir, fileName)
	f, err := os.Create(filePath)
	if err != nil {
		return fmt.Errorf("can't create new file: %w", err)
	}
	defer func(f *os.File) {
		err := f.Close()
		if err != nil {
			fw.logger.Warnf("Failed to close the file %s", filePath)
		}
	}(f)
	fw.logger.Debugf("Create a trace file at [%s]", filePath)
	bytes, err := json.Marshal(group)
	if err != nil {
		return fmt.Errorf("can't marshal DataGroup: %w", err)
	}
	_, err = f.Write(bytes)
	return err
}

func (fw *fileWriter) rotateFiles(baseDir string) error {
	// No constrains set
	if fw.config.MaxFileCountEachProcess <= 0 {
		return nil
	}
	// Get all files path
	toBeRotated, err := getDirEntryInTimeOrder(baseDir)
	if err != nil {
		return fmt.Errorf("can't get files list: %w", err)
	}
	// No need to rotate
	if len(toBeRotated) < fw.config.MaxFileCountEachProcess {
		return nil
	}
	// Remove the older files and remove half of them one time to decrease the frequency
	// of deleting files. Note this is different from rotating log files. We could delete
	// one file at a time for log files because the action "rotate" is in a low frequency
	// in that case.
	toBeRotated = toBeRotated[:len(toBeRotated)-fw.config.MaxFileCountEachProcess/2+1]
	// Remove the stale files asynchronously
	go func() {
		for _, dirEntry := range toBeRotated {
			_ = os.Remove(filepath.Join(baseDir, dirEntry.Name()))
			fw.logger.Infof("Rotate trace files [%s]", dirEntry.Name())
		}
	}()
	return nil
}

// getDirEntryInTimeOrder returns the directory entries slice in chronological order.
// The result files are sorted based on their modification time.
func getDirEntryInTimeOrder(path string) ([]os.DirEntry, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer func(f *os.File) {
		err := f.Close()
		if err != nil {
			return
		}
	}(f)
	dirs, err := f.ReadDir(-1)
	// Sort the files based on their modification time. We don't sort them based on the
	// timestamp in the file name because they are similar but the latter one costs more CPU
	// considering that we have to split the file name first.
	sort.Slice(dirs, func(i, j int) bool {
		fileInfoA, err := dirs[i].Info()
		if err != nil {
			return false
		}
		fileInfoB, err := dirs[j].Info()
		if err != nil {
			return false
		}
		return fileInfoA.ModTime().Before(fileInfoB.ModTime())
	})
	return dirs, err
}

const dividingLine = "\n------\n"

func (fw *fileWriter) writeCpuEvents(group *model.DataGroup) {
	traceTimestamp := group.Labels.GetIntValue(constlabels.Timestamp)
	pathElements := filepathhelper.GetFilePathElements(group, uint64(traceTimestamp))
	baseDir := fw.pidFilePath(pathElements.WorkloadName, pathElements.PodName, pathElements.ContainerName, pathElements.Pid)
	fileName := getFileName(pathElements.Protocol, pathElements.ContentKey, pathElements.Timestamp, pathElements.IsServer)
	filePath := filepath.Join(baseDir, fileName)
	f, err := os.OpenFile(filePath, os.O_WRONLY|os.O_APPEND, 0)
	if err != nil {
		fw.logger.Infof("Couldn't open the trace file %s when append CpuEvents: %v. "+
			"Maybe the file has been rotated.", filePath, err)
		return
	}
	defer func(f *os.File) {
		err := f.Close()
		if err != nil {
			fw.logger.Warnf("Failed to close the file %s, %v", filePath, err)
		}
	}(f)
	_, err = f.Write([]byte(dividingLine))
	if err != nil {
		fw.logger.Errorf("Failed to append CpuEvents to the file %s: %v", filePath, err)
		return
	}
	eventsBytes, _ := json.Marshal(group)
	_, err = f.Write(eventsBytes)
	if err != nil {
		fw.logger.Errorf("Failed to append CpuEvents to the file %s: %v", filePath, err)
		return
	}
	fw.logger.Debugf("Write CpuEvents to trace files [%s]", filePath)
}

func (fw *fileWriter) name() string {
	return storageFile
}

File: refactor/camera-agent-main/src/camera_exporter/file_writer.rs
// use std::fs::{self, DirEntry, OpenOptions};
// use std::io::{self, Write};
// use std::path::{Path, PathBuf};
// use std::time::{SystemTime, UNIX_EPOCH};

// use super::config::FileConfig;
// use crate::model::{consts, Trace, TraceProfiling};
// pub(crate) struct FileWriter {
//     config: FileConfig,
// }
// const DIVIDING_LINE: &str = "\n------\n";

// impl FileWriter {
//     fn new(config: FileConfig) -> io::Result<Self> {
//         Ok(FileWriter { config })
//     }

//     fn pid_file_path(
//         &self,
//         workload_name: &str,
//         pod_name: &str,
//         container_name: &str,
//         pid: i64,
//     ) -> PathBuf {
//         let pid_str = pid.to_string();
//         let mut path = PathBuf::new();
//         path.push(&self.config.storage_path);
//         path.push(workload_name);
//         path.push(pod_name);
//         path.push(container_name);
//         path.push(pid_str);
//         path
//     }

//     fn get_file_name(protocol: &str, content_key: &str, timestamp: u64, is_server: bool) -> String {
//         let mut file_name = String::new();
//         file_name.push_str(protocol);
//         file_name.push_str("_");
//         file_name.push_str(content_key);
//         file_name.push_str("_");
//         file_name.push_str(&timestamp.to_string());
//         if is_server {
//             file_name.push_str("_server");
//         } else {
//             file_name.push_str("_client");
//         }
//         file_name
//     }

//     fn write_trace(
//         &self,
//         group: model::Trace,
//     ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
//         let trace_timestamp = group.labels.get_int_value(consts::TIMESTAMP);
//         let path_elements = get_file_path_elements(group, trace_timestamp as u64);
//         let base_dir = self.pid_file_path(
//             &path_elements.workload_name,
//             &path_elements.pod_name,
//             &path_elements.container_name,
//             path_elements.pid,
//         );
//         let file_name = Self::get_file_name(
//             &path_elements.protocol,
//             &path_elements.content_key,
//             path_elements.timestamp,
//             path_elements.is_server,
//         );
//         let file_path = base_dir.join(&file_name);
//         let mut file = OpenOptions::new()
//             .create(true)
//             .append(true)
//             .open(&file_path)?;
//         writeln!(file, "------")?;
//         let events_bytes = serde_json::to_vec(group)?;
//         file.write_all(&events_bytes)?;
//         Ok(())
//     }

//     fn write_profiling(
//         &self,
//         group: model::TraceProfiling,
//     ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
//         let trace_timestamp = group.labels.get_int_value(consts::TIMESTAMP);
//         let path_elements = get_file_path_elements(group, trace_timestamp as u64);
//         let base_dir = self.pid_file_path(
//             &path_elements.workload_name,
//             &path_elements.pod_name,
//             &path_elements.container_name,
//             path_elements.pid,
//         );
//         let file_name = Self::get_file_name(
//             &path_elements.protocol,
//             &path_elements.content_key,
//             path_elements.timestamp,
//             path_elements.is_server,
//         );
//         let file_path = base_dir.join(&file_name);
//         let mut file = OpenOptions::new()
//             .create(true)
//             .append(true)
//             .open(&file_path)?;
//         writeln!(file, "{}", DIVIDING_LINE)?;
//         let events_bytes = serde_json::to_vec(group)?;
//         file.write_all(&events_bytes)?;
//         Ok(())
//     }

//     fn rotate_files(&self, base_dir: &Path) -> io::Result<()> {
//         let mut entries = fs::read_dir(base_dir)?
//             .map(|res| res.map(|e| e.path()))
//             .collect::<Result<Vec<_>, io::Error>>()?;
//         entries.sort_by_key(|entry| entry.metadata().unwrap().modified().unwrap());
//         let max_files = self.config.max_files;
//         if entries.len() > max_files {
//             let num_files_to_delete = entries.len() - max_files;
//             for entry in entries.iter().take(num_files_to_delete) {
//                 fs::remove_file(entry)?;
//             }
//         }
//         Ok(())
//     }

//     fn get_dir_entry_in_time_order(path: &Path) -> io::Result<Vec<DirEntry>> {
//         let mut entries = fs::read_dir(path)?.collect::<Result<Vec<_>, io::Error>>()?;
//         entries.sort_by_key(|entry| entry.metadata().unwrap().modified().unwrap());
//         Ok(entries)
//     }

//     fn name(&self) -> &str {
//         config::STORAGE_FILE
//     }
// }

// fn get_date_string(timestamp: i64) -> String {
//     let time_unix = UNIX_EPOCH + std::time::Duration::from_nanos(timestamp as u64);
//     let time = time_unix.duration_since(SystemTime::UNIX_EPOCH).unwrap();
//     let date = time.as_secs();
//     let nano = time.subsec_nanos();
//     let year = (date / 31536000) as i32;
//     let month = ((date - (year as u64 * 31536000)) / 2592000) as u32;
//     let day = ((date - (year as u64 * 31536000) - (month as u64 * 2592000)) / 86400) as u32;
//     let hour = ((date - (year as u64 * 31536000) - (month as u64 * 2592000) - (day as u64 * 86400))
//         / 3600) as u32;
//     let minute = ((date
//         - (year as u64 * 31536000)
//         - (month as u64 * 2592000)
//         - (day as u64 * 86400)
//         - (hour as u64 * 3600))
//         / 60) as u32;
//     let second = date
//         - (year as u64 * 31536000)
//         - (month as u64 * 2592000)
//         - (day as u64 * 86400)
//         - (hour as u64 * 3600)
//         - (minute as u64 * 60);
//     return format!(
//         "{}{}{}{}{}{}.{}",
//         date_to_string(year),
//         date_to_string(month),
//         date_to_string(day),
//         date_to_string(hour),
//         date_to_string(minute),
//         date_to_string(second),
//         nano
//     );
// }

// fn date_to_string(date: u32) -> String {
//     if date >= 0 && date <= 9 {
//         return format!("0{}", date);
//     } else {
//         return format!("{}", date);
//     }
// }

// pub struct FilePathElements {
//     pub workload_name: String,
//     pub pod_name: String,
//     pub container_name: String,
//     pub pid: i64,
//     pub is_server: bool,
//     pub protocol: String,
//     pub timestamp: u64,
//     pub content_key: String,
// }

// pub fn get_file_path_elements(group: &TraceProfiling, timestamp: u64) -> FilePathElements {
//     FilePathElements {
//         workload_name: group.workload_name,
//         pod_name: group.pod_name,
//         container_name: group.container_name,
//         pid: group.pid,
//         is_server: group.is_server,
//         protocol: group.protocol,
//         timestamp: group.timestamp,
//         content_key: group.content_key,
//     }
// }
